python-frontend/src/main/java/org/sonar/python/api/IPythonGrammarBuilder.java

```diff
@@ -40,20 +45,32 @@
 import static org.sonar.python.api.PythonGrammar.RAISE_STMT;
 import static org.sonar.python.api.PythonGrammar.RETURN_STMT;
 import static org.sonar.python.api.PythonGrammar.SIMPLE_STMT;
+import static org.sonar.python.api.PythonGrammar.STATEMENT;
 import static org.sonar.python.api.PythonGrammar.TESTLIST_STAR_EXPR;
 import static org.sonar.python.api.PythonGrammar.YIELD_EXPR;
 import static org.sonar.python.api.PythonGrammar.YIELD_STMT;
+import static org.sonar.python.api.PythonTokenType.IPYNB_CELL_DELIMITER;
 import static org.sonar.python.api.PythonTokenType.NEWLINE;
 
 public class IPythonGrammarBuilder extends PythonGrammarBuilder {
 
   @Override
-  protected void setupRules(LexerfulGrammarBuilder b) {
-    lineMagic(b);
-    super.setupRules(b);
+  public Grammar create() {
+    LexerfulGrammarBuilder b = LexerfulGrammarBuilder.create();
+
+    b.rule(FILE_INPUT).is(b.zeroOrMore(b.firstOf(CELL, MAGIC_CELL, NEWLINE, IPYNB_CELL_DELIMITER)), EOF);
+
+    setupIPythonNotebooksRules(b);
+    setupCommonRules(b);
+
+    b.setRootRule(FILE_INPUT);
+    return b.buildWithMemoizationOfMatchesForAllRules();
   }
 
-  protected void lineMagic(LexerfulGrammarBuilder b) {
+  protected void setupIPythonNotebooksRules(LexerfulGrammarBuilder b) {
```

Comment:
in another PR there is a separate method for another IPython-specific grammar, I think is better to keep methods responsibility more atomic instead of one method for everything

Reply:
I kind of wanted to group IPython rules to a method, similar to how we already group rules for `statements`, `expressions`, etc.. in one method per category rather than one method per rule.
Although it's true that here, the various elements from the `IPython` category could be broken down further (in cells, statements and expressions for example), but I don't think we'll have so much more rules, and that should not affect parsing results either.